### PR TITLE
SONARUNNER-148 SONARUNNER-149 TLS1.2 on Java7 + do not ignore security

### PR DESCRIPTION
**Pull request by simonbrandhof**
SONARUNNER-148 SONARUNNER-149 TLS1.2 on Java7 + do not ignore security

### CODE REVIEW

File: sonar-runner-api/src/main/java/org/sonar/runner/impl/Tls12Java7SocketFactory.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import javax.net.ssl.SSLSocketFactory;
+
+/**
+ * {@link SSLSocketFactory} which enables all the versions of TLS. This is required
+ * to support TLSv1.2 on Java 7. Note that Java 8 supports TLSv1.2 natively, without
+ * any configuration
+ */
+public class Tls12Java7SocketFactory extends SSLSocketFactory {
+
+  public static final String[] TLS_PROTOCOLS = new String[]{"TLSv1", "TLSv1.1", "TLSv1.2"};
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/impl/Tls12Java7SocketFactory.java]: CRITICAL Make this array "private". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1873)
CRITICAL Make this member "protected". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2386)

File: sonar-runner-api/src/main/java/org/sonar/runner/impl/Tls12Java7SocketFactory.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  @Override
+  public Socket createSocket(InetAddress host, int port, InetAddress localAddress, int localPort) throws IOException {
+    Socket underlyingSocket = delegate.createSocket(host, port, localAddress, localPort);
+    return overrideProtocol(underlyingSocket);
+  }
+
+  /**
+   * Enables TLS v1.0, 1.1 and 1.2 on the socket
+   */
+  private Socket overrideProtocol(Socket socket) {
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/impl/Tls12Java7SocketFactory.java]: MAJOR Make "overrideProtocol" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: sonar-runner-api/src/main/java/org/sonar/runner/impl/ServerConnection.java [hunk 3]
```diff
@@ -190,29 +171,20 @@ private String getFromCache(String fullUrl) {
...
+   * @throws IllegalStateException if HTTP code is different than 2xx
    */
-  static String getCharsetFromContentType(String contentType) {
-    if (contentType == null) {
-      return null;
-    }
-    Matcher m = CHARSET_PATTERN.matcher(contentType);
-    if (m.find()) {
-      return m.group(1).trim().toUpperCase();
+  private ResponseBody callUrl(String url) throws IOException, IllegalStateException {
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/impl/ServerConnection.java]: MAJOR Remove the declaration of thrown exception 'java.lang.IllegalStateException' which is a runtime exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

File: sonar-runner-api/src/main/java/org/sonar/runner/impl/ServerConnection.java [hunk 4]
```diff
@@ -19,169 +19,150 @@
...
+   * @throws IOException           if connectivity problem or timeout (network) or IO error (when writing to file)
+   * @throws IllegalStateException if HTTP response code is different than 2xx
+   */
+  private String downloadString(String url, boolean saveCache) throws IOException {
+    logger.debug(format("Download: %s", url));
+    ResponseBody responseBody = callUrl(url);
+    String content = responseBody.string();
+    if (saveCache) {
+      try {
+        wsCache.put(url, content.getBytes());
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/impl/ServerConnection.java]: MAJOR Remove this use of "getBytes" [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1943)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* CRITICAL 2 critical
* MAJOR 3 major

Watch the comments in this conversation to review them.

File: sonar-runner-api/src/main/java/org/sonar/runner/impl/Tls12Java7SocketFactory.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * {@link SSLSocketFactory} which enables all the versions of TLS. This is required
+ * to support TLSv1.2 on Java 7. Note that Java 8 supports TLSv1.2 natively, without
+ * any configuration
+ */
+public class Tls12Java7SocketFactory extends SSLSocketFactory {
+
+  // visible for testing
+  static final String[] TLS_PROTOCOLS = new String[]{"TLSv1", "TLSv1.1", "TLSv1.2"};
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/impl/Tls12Java7SocketFactory.java]: CRITICAL Make this array "private". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1873)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

**Comment by dbmeneses** [comment]
LGTM

**Pull request merged**